src/sources/socket/udp.rs

```diff
@@ -121,37 +122,52 @@ pub fn udp(
                     emit!(BytesReceived { byte_size, protocol: "udp" });
 
                     let payload = buf.split_to(byte_size);
+                    let truncated = byte_size == max_length + 1;
 
-                    let mut stream = FramedRead::new(payload.as_ref(), decoder.clone());
+                    let mut stream = FramedRead::new(payload.as_ref(), decoder.clone()).peekable();
 
                     while let Some(result) = stream.next().await {
+                        let last = Pin::new(&mut stream).peek().await.is_none();
                         match result {
                             Ok((mut events, _byte_size)) => {
-                                let count = events.len();
-                                emit!(SocketEventsReceived {
-                                    mode: SocketMode::Udp,
-                                    byte_size: events.size_of(),
-                                    count,
-                                });
-
-                                let now = Utc::now();
-
-                                for event in &mut events {
-                                    if let Event::Log(ref mut log) = event {
-                                        log.try_insert(log_schema().source_type_key(), Bytes::from("socket"));
-                                        log.try_insert(log_schema().timestamp_key(), now);
-                                        log.try_insert(host_key.as_str(), address.to_string());
-                                    }
+
+                                if last && truncated {
+                                    // The last event in this payload was truncated, so we want to drop it.
+                                    let _ = events.pop();
+                                    warn!(
+                                        message = "Discarding frame larger than max_length.",
+                                        max_length = max_length,
+                                        internal_log_rate_secs = 30
+                                    );
```

Comment:
I think this should emit an internal event with the appropriate metrics per https://github.com/vectordotdev/vector/blob/master/docs/specs/component.md#error

Reply:
There are some problems with this.  The [TCP source](https://github.com/vectordotdev/vector/blob/master/lib/codecs/src/decoding/framing/character_delimited.rs#L95-L100) and probably others also doesn't emit an internal event.  So this remains consistent with that. 

It is not completely straightforward to change the TCP source to emit an internal event, since the decoding is in a separate library and the code for events are in the main Vector project - so this would involve some refactoring to enable this.

I think we should keep it like this and have an issue to sort out the events for everything in one go?

Reply:
Aha, I didn't realize that. It seems like it was an oversight in #11124 then. 👍 to spinning it off into its own issue.